src/crimson/net/SocketMessenger.h

```diff
@@ -63,7 +64,10 @@ class SocketMessenger final : public Messenger {
   seastar::future<> try_bind(const entity_addrvec_t& addr,
                              uint32_t min_port, uint32_t max_port) override;
 
-  seastar::future<> start(Dispatcher *dispatcher) override;
+  seastar::future<> start(ChainedDispatchersRef& dispatchers) override;
```

Comment:
Um...this is actually a little complicated.

First, we have to use slist with "boost::intrusive::linear<true>", cause otherwise it would triggered exceptions when removing hooks while there are other iterations going on;
Second, when using slist with "boost::intrusive::linear<true>" when can't determine a hook is lined if that hook is at the tail of the slist; 
Third, there are scenarios like mkfs where no dispatchers are added to the slist, which make it necessary to check whether a component is linked when shutting down that component;

You can see that the 2nd point is conflicted with the 3rd point, so I made the existence chained_dispatcher a indication of whether dispatchers are added. So it seems necessary that we added it to the messenger when starting it.

Reply:
> You can see that the 2nd point is conflicted with the 3rd point, so I made the existence of the chained_dispatcher an indication of whether dispatchers are added. So it seems necessary that we add it to the messenger when starting it.

I think it is orthogonal, it's possible to address it inside `remove_dispatcher()`. My concern is that it makes inlining `ChainedDispatchers` impossible and leads to side-effects as said in https://github.com/ceph/ceph/pull/34492#discussion_r427390403. Basically I mean it could be simpler if the possible states for messenger is just 0~n dispatchers, and other sub-components doesn't need to worry that they are holding a different set of dispatchers.

Reply:
There's a restriction from intrusive slist, see https://github.com/ceph/ceph/pull/34492/commits/51a4b39fcfe2d627572935fe403eaa2cf5338858#r428547396

One consequence is that  `add_dispatcher()` and `remove_dispatcher()` can impact other messengers.